### ISSUE TITLE
Handle program runtime error handling

### ISSUE THREAD

**Issue by aeyakovenko** [solana-labs/solana#1335]
This is related to #1157 

Programs that execute need to pay a transaction fee, so that transaction needs to be encoded into the ledger.  Otherwise the network can be spammed.  Runtime errors are an important control flow mechansim
* developers need to be able to shortcut and fail the transaction programatically
* the runtime needs to communicate the error back up the stack to external apis

The proposed solution is as follows:

1. programs are scheduled to execute if there are enough tokens to pay the execution fee
2. programs can fail and return a runtime error
3. the runtime error is stored in the recent signature table
4. the fee is withdrawn from the account
5. the memory changes are not committed
6. the transaction is committed to the ledger

@mvines  @garious (appreciate your feedback!)

**Comment by mvines** [comment]
Let's define the width of the runtime error code (16bits?), lower 8 bits for system error codes and high 8 bits for program-specific error codes?

**Comment by aeyakovenko** [comment]
@mvines the signature's error state would also be storing the program index that caused the runtime error.
```
Result<()>
Ok/Err is 8 bytes afaik
BankError is 8 bytes as well
RuntimeError(index: u8, errorcode: u8)
```
it will probably allocate at min 24 bytes for this thing. unless we just store it as bincode serialized.

**Comment by aeyakovenko** [comment]
@mvines rust's enums are 8 bytes :), so

**Comment by garious** [comment]
Weird, they must of changed that. Last I checked, Rust enums would only be 8-bytes if you added `CRepr`. Otherwise it's at the compiler's discretion and I thought the compiler used just 1 byte for an enum until it got past 256 items.

**Comment by aeyakovenko** [comment]
https://docs.google.com/presentation/d/1q-c7UAyrUlM-eZyTo1pd8SZ0qwA_wYxmPZVOQkoDmH4/mobilepresent?slide=id.p

**Comment by garious** [comment]
https://d3m3vilurr.gitbooks.io/the-unsafe-rust-programming-language/content/data.html

**Comment by aeyakovenko** [comment]
@garious. How old is that book? “Note: this is a hypothetical optimization that is not yet implemented in Rust 1.0.0”

**Comment by garious** [comment]
Programming in Rust, "Enums in Memory," page 215:

> In memory, enums with data are stored as a small integer *tag,* plus enough memory to hold all the fields of the largest variant. The tag field is for Rust's internal use. It ells which constructor created the value, and therefore which fields it has.

And then it goes on to show a picture of the *tag* being stored in 1 byte.
